lib/HttpClient.ts

```diff
@@ -52,6 +53,28 @@ export class HttpClientResponse implements ifm.IHttpClientResponse {
 
     public message: http.IncomingMessage;
     readBody(): Promise<string> {
+        // Extract Encoding from header: 'content-encoding'
+        // Match `gzip`, `gzip, deflate` variations of GZIP encoding
+        const contentEncoding: string = this.message.headers['content-encoding'] || '';
+        const isGzippedEncoded: boolean = new RegExp('(gzip$)|(gzip, *deflate)').test(contentEncoding);
+
+        if (isGzippedEncoded) { //Process GZIPPED Response
+            return new Promise<string>(async (resolve, reject) => {
+                let buffer: Buffer = Buffer.alloc(0);
+                const encodingCharset = util.obtainContentCharset(this);
+
+                this.message.on('data', function(data: string|Buffer) {
+                    const chunk = (typeof data === 'string') ? Buffer.from(data, encodingCharset) : data;
+                    buffer = Buffer.concat([buffer, chunk]);
+                }).on('end', async function() {
+                    const output = await util.decompressGzippedContent(buffer, encodingCharset);
+                    resolve(output);
+                }).on('error', function(err) {
+                    reject(err);
+                });
+            });
+        }
```

Comment:
Nit: We can probably consolidate this into a single return. Would be good to get your buffer/string detection in both anyways. So the single returned promise would look something like:

```
return new Promise<string>(async (resolve, reject) => {
    let buffer: Buffer = Buffer.alloc(0);
    const encodingCharset = util.obtainContentCharset(this);

    this.message.on('data', function(data: string|Buffer) {
        const chunk = (typeof data === 'string') ? Buffer.from(data, encodingCharset) : data;
        buffer = Buffer.concat([buffer, chunk]);
    }).on('end', async function() {
        if (isGzippedEncoded) {
            resolve(await util.decompressGzippedContent(buffer, encodingCharset));
        } else {
            resolve(buffer.toString());
        }
    }).on('error', function(err) {
        reject(err);
    });
});
```

Thoughts?

Reply:
Yeah, Totally agree here too! I did think of producing a sole return statement with joining code, don't know why left it that way. Event the inner `if-else` block within the `on('end')` event can be consolidated to a sole if statement to handle gzipped content, OR ELSE return the stringified JSON buffer.

After re-arranging, sorting-out the code, turns out to be something as:
```
readBody(): Promise<string> {
        return new Promise<string>(async (resolve, reject) => {
            let buffer: Buffer = Buffer.alloc(0);
            const encodingCharset = util.obtainContentCharset(this);

            // Extract Encoding from header: 'content-encoding'
            // Match `gzip`, `gzip, deflate` variations of GZIP encoding
            const contentEncoding: string = this.message.headers['content-encoding'] || '';
            const isGzippedEncoded: boolean = new RegExp('(gzip$)|(gzip, *deflate)').test(contentEncoding);

            this.message.on('data', function(data: string|Buffer) {
                const chunk = (typeof data === 'string') ? Buffer.from(data, encodingCharset) : data;
                buffer = Buffer.concat([buffer, chunk]);
            }).on('end', async function() {
                if (isGzippedEncoded) {
                    const gunzippedBody = await util.decompressGzippedContent(buffer, encodingCharset);
                    resolve(gunzippedBody);
                }

                resolve(buffer.toString(encodingCharset));
            }).on('error', function(err) {
                reject(err);
            });
        });
    }
```

Reply:
Yep, that looks great!